kroxylicious-filters/kroxylicious-encryption/src/main/java/io/kroxylicious/filter/encryption/inband/InBandDecryptionManager.java

```diff
@@ -49,17 +50,67 @@
  */
 public class InBandDecryptionManager<K, E> implements DecryptionManager {
 
-    private final AsyncLoadingCache<E, AesGcmEncryptor> decryptorCache;
-    private final Kms<K, E> kms;
-    private final Serde<E> edekSerde;
+    private static final Logger LOGGER = LoggerFactory.getLogger(InBandDecryptionManager.class);
+
+    public static final int NO_MAX_CACHE_SIZE = -1;
+
+    private final AsyncLoadingCache<CacheKey<E>, Dek<E>> decryptorCache;
+    private final DekManager<K, E> dekManager;
     private final FilterThreadExecutor filterThreadExecutor;
 
-    public InBandDecryptionManager(Kms<K, E> kms, FilterThreadExecutor filterThreadExecutor) {
-        this.kms = kms;
-        this.edekSerde = kms.edekSerde();
+    private record CacheKey<E>(
+                               @Nullable CipherSpec cipherSpec,
+                               @Nullable E edek) {
+        @SuppressWarnings("rawtypes")
+        private static final CacheKey NONE = new CacheKey(null, null);
+
+        @SuppressWarnings("unchecked")
+        static <E> CacheKey<E> none() {
+            return NONE;
+        }
+
+        public boolean isNone() {
+            return cipherSpec == null || edek == null;
+        }
+    }
+
+    public InBandDecryptionManager(DekManager<K, E> dekManager,
+                                   @Nullable FilterThreadExecutor filterThreadExecutor,
+                                   @Nullable Executor dekCacheExecutor,
+                                   int dekCacheMaxItems) {
+        this.dekManager = dekManager;
         this.filterThreadExecutor = filterThreadExecutor;
-        this.decryptorCache = Caffeine.newBuilder()
-                .buildAsync((edek, executor) -> makeDecryptor(edek));
+        Caffeine<Object, Object> cache = Caffeine.<CacheKey<E>, Dek<E>> newBuilder();
+        if (dekCacheMaxItems != NO_MAX_CACHE_SIZE) {
+            cache = cache.maximumSize(dekCacheMaxItems);
+        }
+        if (dekCacheExecutor != null) {
+            cache = cache.executor(dekCacheExecutor);
+        }
+        // TODO support batched resolution in the DekManager and also in the KMS
+        this.decryptorCache = cache
```

Comment:
I opened https://github.com/kroxylicious/kroxylicious/issues/976 for supporting batched loading into the cache.